macaroonbakery/macaroon.py

```diff
@@ -157,7 +167,38 @@ def _new_caveat_id(self, base):
         @param base string
         @return string
         '''
-        raise NotImplementedError
+        id = bytearray()
+        if len(base) > 0:
+            id.append(base)
+        else:
+            # Add a version byte to the caveat id.Technically
+            # this is unnecessary as the caveat-decoding logic
+            # that looks at versions should never see this id,
+            # but if the caveat payload isn't provided with the
+            # payload, having this version gives a strong indication
+            # that the payload has been omitted so we can produce
+            # a better error for the user.
+            id.append(BAKERY_V3)
+        id.append(0)
+        # Iterate through integers looking for one that isn't already used,
+        # starting from n so that if everyone is using this same algorithm,
+        # we'll only perform one iteration.
+        i = len(self._caveat_data)
+        caveats = self._macaroon.caveats
+        while True:
+            temp = bytearray()
+            # We append a varint to the end of the id and assume that
+            # any client that's created the id that we're using as a base
+            # is using similar conventions - in the worst case they might
+            # end up with a duplicate third party caveat id and thus create
+            # a macaroon that cannot be discharged.
+            temp[:] = id
+            codec.encode_uvarint(i, temp)
+            for cav in caveats:
+                if cav.verification_id is not None and cav.caveat_id == temp:
+                    i += 1
+                    break
+            return temp.decode('utf-8')
```

Comment:
This doesn't seem right to me, as not all uvarints can be encoded as utf-8.
Shouldn't this be bytes(temp) ?

Reply:
is there a mistake in the go implementation as the only place it is used is here https://github.com/go-macaroon-bakery/macaroon-bakery/blob/v2-unstable/bakery/macaroon.go#L303 where we just treansform it to string directly ? @rogpeppe

Reply:
fixed